source/extensions/http/cache/file_system_http_cache/file_system_http_cache.cc

```diff
@@ -388,6 +404,108 @@ void FileSystemHttpCache::trackFileRemoved(uint64_t file_size) {
   stats_.size_bytes_.set(size_bytes_);
 }
 
+bool FileSystemHttpCache::needsEviction() const {
+  if (config().has_max_cache_size_bytes() &&
+      size_bytes_ > config().max_cache_size_bytes().value()) {
+    return true;
+  }
+  if (config().has_max_cache_entry_count() &&
+      size_count_ > config().max_cache_entry_count().value()) {
+    return true;
+  }
+  return false;
+}
+
+bool FileSystemHttpCache::needsEviction(uint64_t& size_to_evict_out,
+                                        uint64_t& count_to_evict_out) const {
+  if (config().has_max_cache_size_bytes()) {
+    // capture a value from the atomic because we're going to use it twice and don't want our
+    // value to change in between.
+    uint64_t seen_size = size_bytes_;
+    size_to_evict_out = seen_size > config().max_cache_size_bytes().value()
+                            ? seen_size - config().max_cache_size_bytes().value()
+                            : 0;
+  } else {
+    size_to_evict_out = 0;
+  }
+  if (config().has_max_cache_entry_count()) {
+    // capture a value from the atomic because we're going to use it twice and don't want our
+    // value to change in between.
+    uint64_t seen_count = size_count_;
+    count_to_evict_out = seen_count > config().max_cache_entry_count().value()
+                             ? seen_count - config().max_cache_entry_count().value()
+                             : 0;
+  } else {
+    count_to_evict_out = 0;
+  }
+  return size_to_evict_out > 0 || count_to_evict_out > 0;
+}
+
+void FileSystemHttpCache::maybeEvict() {
+  uint64_t size_to_evict, count_to_evict;
+  if (!needsEviction(size_to_evict, count_to_evict)) {
```

Comment:
I'm wondering if we need to re-call this here. I don't know if it's possible to call maybeEvict() now unless an eviction is needed. And then you may as well do the directory walk and get the more exact numbers.

Reply:
You'd have to do the directory walk twice in that case, once to find the size and then again to select the files to evict now that you know how much you need to evict.
Also, the pre-check is necessary because the eviction thread has a shot at each cache when woken, so if there's more than one cache and only one needs evicting, there's no need to scan the others.

Reply:
I don't see how to get around it unless you trust your cached value (which seems unwise if there are two processes operating in the same directory).

Reply:
I think it would be fine to trust the cached values really, but okay, I've redone it so it doesn't trust the cached values, still only does one pass, and now doesn't block for long, at the small cost of putting all the cache filenames in memory briefly during the eviction pass. (Still doing a pre-check to avoid running eviction over cache instances that don't need it.)

Reply:
(And this should be fine if two instances run at the same time since they'll agree to keep the same files, worst case it would give them a misleading estimate of their cache size afterwards so they might run another eviction pass before it's necessary.)